### ISSUE TITLE
Certificate issues with multiple subdomains

### ISSUE THREAD

**Issue by kevinvdburgt** [evertramos/docker-compose-letsencrypt-nginx-proxy-companion#69]
# Error description:
The reverse proxy gives the wrong certificate when using subdomains of the same domain. For example, i have foobar1.example.com and foobar2.example.com. When visiting foorbar2.example.com it gives a certificate error that it serves the certificate form foobar1 to the foobar2 subdomain.
<img width="1277" alt="screen shot 2018-05-28 at 16 27 52" src="https://user-images.githubusercontent.com/1987802/40619114-1c7617b6-6294-11e8-906d-c297699385df.png">

# Steps to reproduce
Created 3 sites using subdomains:
- foobar1.example.com
- foobar2.example.com
- foobar3.example.com

(restart the server a few times until it happens, sometimes it happens without restarting at all)

visit foobar2.example.com and it will show the certificate of foobar1.example.com

Used those `docker-compose.yml` config files to test:
```yml
version: '3'

services:
  app:
    image: httpd:alpine
    environment:
      VIRTUAL_HOST: site-a.node3.docker.zdev.com
      VIRTUAL_PORT: 80
      LETSENCRYPT_HOST: site-a.node3.docker.zdev.com
      LETSENCRYPT_EMAIL: info@zdev.com
    restart: always
    networks:
      - webproxy

networks:
  webproxy:
    external:
      name: webproxy
```

```yml
version: '3'

services:
  app:
    image: httpd:alpine
    environment:
      VIRTUAL_HOST: site-b.node3.docker.zdev.com
      VIRTUAL_PORT: 80
      LETSENCRYPT_HOST: site-b.node3.docker.zdev.com
      LETSENCRYPT_EMAIL: info@zdev.com
    restart: always
    networks:
      - webproxy

networks:
  webproxy:
    external:
      name: webproxy
```

```yml
version: '3'

services:
  app:
    image: httpd:alpine
    environment:
      VIRTUAL_HOST: site-c.node3.docker.zdev.com
      VIRTUAL_PORT: 80
      LETSENCRYPT_HOST: site-c.node3.docker.zdev.com
      LETSENCRYPT_EMAIL: info@zdev.com
    restart: always
    networks:
      - webproxy

networks:
  webproxy:
    external:
      name: webproxy
```

# Expected result
That all of the domains uses the correct certificates.

# Error result
Could not find any error in the logs.

# Comments
It works fine for domains like foobar1.com, foobar2.com, foobar3.com .. but when using subdomains of the same domain it will fail after a period of time (mostly after restarting the server).

Sometimes it just works for days, and when the server (CoreOS) restarts itself for updates and stuff it happens to serve the wrong certificates again.

**Comment by kevinvdburgt** [comment]
Not 100% sure about this, but:

I might have an idea of what is happening, everything runs fine until a restart of docker (e.g. system reboot). When that happens and docker is going to auto start the services again it might happen that site (lets call it foo.bar.com) finishes starting before this project (the webproxy) and another site (lets call it hello.world.com) finishes as last, so the startup order is:

1. Project foo.bar.com
2. Webproxy (this project)
3. Project hello.world.com

Which means that hello.world.com has been seen by the webproxy and can do the thing for createing the vhost config. However, when the project foo.bar.com starts, there is no way for the webproxy to see that project is being started.

**Comment by evertramos** [comment]
Sorry I could not check this earlier... I will try to check it over the weekend. Thanks for the feedback.

**Comment by magneticflux-** [comment]
This has happened to me as well, and it seems related to #73.

**Comment by philipkobernik** [comment]
@magneticflux- I closed #73, error must be in my config because a stripped down config to test site.com and staging.site.com containers worked flawlessly. See demo https://github.com/philipkobernik/docker-compose-letsencrypt-nginx-proxy-demo

**Comment by magneticflux-** [comment]
I haven't had any more certificate mixups, however I have had a recurring 502 error problem. After some experimentation, I found that a reliable way to trigger a 502 Bad Gateway error was to attempt to connect many times to the proxied container while it was starting up. This causes the `nginx-web` container to mark the upstream as "down", and so it ignores it for a while.

**Comment by evertramos** [comment]
Hello @kevinvdburgt and @magneticflux-,

Sorry my late reply... but here it goes some comments that might be helpful.

**restarting a server multiple times** sounds weird... since we normally do not restart a Linux Server... normally! But in case its needed everything should be working just fine once each time the webproxy starts its service it reads from the docker socket and get the containers certificates in order to update the nginx conf file. 

I have tested today restarting the server, changed the ip address of the containers, networks and everything just worked fine.

Here are the links which I tested and in ALL the test ti served the correct certificate:

https://iots.dockerplay.com
https://oh.dockerplay.com

Besides that I would suggest you run the `start.sh` script in order to update the containers and the nginx.tmpl file, it might fix any bugs that might happened during your tests.

If you still get this same problem, please reopen this issue and let me know.

**Comment by ablanchard** [comment]
Hi, 
First thanks you for this awesome project
I fall into the same problem lately.
I configured a wordpress site (main site) using your repo `docker-wordpress-letsencrypt` in `.env` I have:
```
DOMAINS=domain.com,www.domain.com
```
Every thing works fine
Later I decided to add another wordpress site (test site), so I clone the repo somewhere else and fill `.env` as:
```
DOMAINS=test.domain.com
```
When I start the test site container, the page is not accessible 
```
Code d’erreur : MOZILLA_PKIX_ERROR_SELF_SIGNED_CERT
```
If I stop the main site container, the test site become accessible.
Stranger... if I restart the main site container, the test site is still accessible but the main site is not.

Hope it add some insights to the problem

**Comment by magneticflux-** [comment]
Since my last comment, I've switched to a Traefik-based reverse proxy and it's been working great so far. If you're running into limitations with this nginx-based reverse proxy, I'd recommend using Traefik for more complex stuff. I've got a half-dozen containers running under different subdomains of my main domain and it's much easier to manage large numbers of proxied containers with labels than using environment variables.
